File: backend/validators.py
# validators.py
# Snippets: normalizador de datas, parser MRZ (passaporte) com checksum,
# validador de receipt USCIS (I-797) e verificador de SSN "válido".
from __future__ import annotations
import re
from datetime import datetime, date

# -----------------------------
# 1) Normalizador de datas
# -----------------------------
# Tenta múltiplos formatos e retorna ISO (YYYY-MM-DD).
# prefer_day_first=True => "12/05/2025" -> 2025-05-12
def normalize_date(s: str, prefer_day_first: bool = True) -> str | None:
    if not s:
        return None
    s = s.strip()
    # Aceitar "D/S" (I-94)
    if s.upper() in {"D/S", "DS"}:
        return "D/S"

    # Formatos comuns
    patterns_df = ["%d/%m/%Y", "%d-%m-%Y", "%d.%m.%Y"]  # day-first
    patterns_mdy = ["%m/%d/%Y", "%m-%d-%Y", "%m.%d.%Y"] # month-first
    patterns_text = ["%b %d, %Y", "%B %d, %Y"]          # "Jan 02, 2025"

    # Heurística: tentar ambos; se prefer_day_first, prioriza DF
    tries = (patterns_df + patterns_mdy + patterns_text) if prefer_day_first \
        else (patterns_mdy + patterns_df + patterns_text)

    for fmt in tries:
        try:
            dt = datetime.strptime(s, fmt).date()
            return dt.isoformat()
        except ValueError:
            continue

    # AAAA-MM-DD já válido?
    try:
        return datetime.strptime(s, "%Y-%m-%d").date().isoformat()
    except ValueError:
        pass
    return None


# -----------------------------
# 2) MRZ (Passaporte TD3) Parser + Checksums
# -----------------------------
# Espera duas linhas TD3 (44 chars cada). Faz parse, valida check digits
# e retorna campos normalizados (numero, DOB, validade, nacionalidade, nome).
_MRZ_WEIGHTS = [7, 3, 1]

def _mrz_char_value(c: str) -> int:
    if c.isdigit():
        return int(c)
    if "A" <= c <= "Z":
        return ord(c) - 55  # A=10 ... Z=35
    if c == "<":
        return 0
    # caracteres inesperados contam como 0
    return 0

def _mrz_checksum(field: str) -> int:
    total = 0
    for i, ch in enumerate(field):
        total += _mrz_char_value(ch) * _MRZ_WEIGHTS[i % 3]
    return total % 10

def parse_mrz_td3(line1: str, line2: str) -> dict | None:
    if len(line1) != 44 or len(line2) != 44:
        return None
    # line1: 0='P', 2-4=issuing country, rest=names
    # line2 fields (positions):
    # 0-8: passport number, 9: check
    # 10-12: nationality (3)
    # 13-18: birth YYMMDD, 19: check
    # 20: sex
    # 21-26: expiry YYMMDD, 27: check
    # 28-42: optional data, 43: composite check
    number = line2[0:9]
    number_c = line2[9]
    nationality = line2[10:13]
    dob = line2[13:19]
    dob_c = line2[19]
    sex = line2[20]
    exp = line2[21:27]
    exp_c = line2[27]
    optional = line2[28:43]
    composite_c = line2[43]

    # Valida checksums
    if _mrz_checksum(number) != (ord(number_c) - ord('0')):
        return None
    if _mrz_checksum(dob) != (ord(dob_c) - ord('0')):
        return None
    if _mrz_checksum(exp) != (ord(exp_c) - ord('0')):
        return None
    # Composite check: concat numero+check + DOB+check + sex + exp+check + optional
    composite_field = number + number_c + nationality + dob + dob_c + sex + exp + exp_c + optional
    if _mrz_checksum(composite_field) != (ord(composite_c) - ord('0')):
        return None

    # Nome (line1): P<ISS<<SURNAME<<GIVEN<NAMES<<<<
    # Extrair após os 5 primeiros chars ("P<XXX")
    raw_names = line1[5:].rstrip("<")
    parts = [p for p in raw_names.split("<<") if p]
    surname = parts[0].replace("<", " ") if len(parts) >= 1 else ""
    given = parts[1].replace("<", " ") if len(parts) >= 2 else ""

    def _yyMMdd_to_iso(yyMMdd: str) -> str | None:
        # Resolve século usando regra simples: YY >= 50 -> 19YY; senão 20YY
        try:
            yy = int(yyMMdd[0:2]); mm = int(yyMMdd[2:4]); dd = int(yyMMdd[4:6])
        except Exception:
            return None
        year = 1900 + yy if yy >= 50 else 2000 + yy
        try:
            return date(year, mm, dd).isoformat()
        except ValueError:
            return None

    return {
        "passport_number": number.replace("<", ""),
        "nationality": nationality,
        "date_of_birth": _yyMMdd_to_iso(dob),
        "expiry_date": _yyMMdd_to_iso(exp),
        "sex": sex,
        "surname": surname.strip(),
        "given_names": given.strip(),
    }


# -----------------------------
# 3) Receipt USCIS (I-797) Validator
# -----------------------------
# Formato: 3 letras + 10 dígitos (ex.: SRC1234567890), com prefixos conhecidos.
_VALID_PREFIXES = {
    "EAC", "WAC", "LIN", "SRC", "MSC", "IOE", "NBC", "NSC", "TSC", "VSC", "YSC"
}

_receipt_rx = re.compile(r"\b([A-Z]{3})(\d{10})\b")

def is_valid_uscis_receipt(s: str) -> bool:
    if not s:
        return False
    m = _receipt_rx.search(s.strip().upper())
    if not m:
        return False
    prefix = m.group(1)
    return prefix in _VALID_PREFIXES


# -----------------------------
# 4) SSN (padrões válidos)
# -----------------------------
# Regras básicas: AAA-GG-SSSS
# - Área AAA != 000, 666, 900-999
# - Grupo GG != 00
# - Série SSSS != 0000
_ssn_rx = re.compile(r"\b(\d{3})-(\d{2})-(\d{4})\b")

def is_plausible_ssn(s: str) -> bool:
    if not s:
        return False
    m = _ssn_rx.search(s.strip())
    if not m:
        return False
    area, group, serial = int(m.group(1)), int(m.group(2)), int(m.group(3))
    if area == 0 or group == 0 or serial == 0:
        return False
    if area == 666 or area >= 900:
        return False
    return True

# -----------------------------
# 5) Validadores adicionais baseados no sistema atual
# -----------------------------

def validate_passport_number_with_nationality(number: str, nationality: str) -> tuple[bool, float, str]:
    """
    Valida número de passaporte baseado na nacionalidade
    Integra com sistema existente
    """
    if not number:
        return False, 0.0, "Passport number is missing"
    
    # Normalize number (remove spaces, uppercase)
    normalized = re.sub(r'[^\w]', '', number.upper())
    
    # Country-specific patterns
    country_patterns = {
        'BR': r'^[A-Z]{2}\d{6}$',  # Brazilian passport
        'BRA': r'^[A-Z]{2}\d{6}$',
        'US': r'^\d{9}$',          # US passport
        'USA': r'^\d{9}$'
    }
    
    # Get country code from nationality
    country_code = None
    if nationality:
        nat_lower = nationality.lower()
        if 'brazil' in nat_lower or 'brasil' in nat_lower:
            country_code = 'BR'
        elif 'united states' in nat_lower or 'american' in nat_lower:
            country_code = 'US'
    
    # Validate against country pattern
    if country_code and country_code in country_patterns:
        pattern = country_patterns[country_code]
        if re.match(pattern, normalized):
            return True, 0.98, "Valid passport number format"
        else:
            return False, 0.3, f"Invalid {country_code} passport number format"
    
    # Generic validation for unknown countries
    if 6 <= len(normalized) <= 12 and normalized.isalnum():
        return True, 0.85, "Valid generic passport number format"
    
    return False, 0.2, "Invalid passport number format"

def validate_date_with_context(date_str: str, field_type: str = 'general') -> tuple[bool, float, str, str]:
    """
    Valida data com contexto específico do campo
    Returns: (is_valid, confidence, normalized_date, error_message)
    """
    if not date_str:
        return False, 0.0, '', 'Date is missing'
    
    # Try to normalize using the robust normalizer
    normalized = normalize_date(date_str.strip())
    
    if not normalized:
        return False, 0.0, '', 'Invalid date format'
    
    if normalized == "D/S":
        return True, 1.0, "D/S", ""  # Valid for I-94
    
    try:
        parsed_date = datetime.strptime(normalized, '%Y-%m-%d')
        current_date = datetime.now()
        
        # Field-specific validation
        if field_type == 'birth_date':
            if parsed_date.year < 1900:
                return False, 0.3, normalized, 'Birth year too old'
            if parsed_date > current_date:
                return False, 0.1, normalized, 'Birth date in future'
            
        elif field_type == 'expiry_date':
            if parsed_date < current_date:
                return False, 0.5, normalized, 'Document has expired'
            
        elif field_type == 'issue_date':
            if parsed_date > current_date:
                return False, 0.3, normalized, 'Issue date in future'
        
        return True, 0.95, normalized, ""
        
    except ValueError:
        return False, 0.0, '', 'Invalid date value'

def extract_and_validate_mrz(document_text: str) -> dict | None:
    """
    Extrai e valida MRZ de texto OCR
    """
    lines = document_text.split('\n')
    
    # Look for two consecutive lines of 44 characters (TD3 format)
    for i in range(len(lines) - 1):
        line1 = lines[i].strip()
        line2 = lines[i + 1].strip()
        
        if len(line1) == 44 and len(line2) == 44:
            # Check if it looks like MRZ (starts with P< for passport)
            if line1.startswith('P<'):
                mrz_data = parse_mrz_td3(line1, line2)
                if mrz_data:
                    return {
                        'mrz_found': True,
                        'mrz_valid': True,
                        'mrz_data': mrz_data,
                        'confidence': 0.99
                    }
    
    return None

# -----------------------------
# 6) Função de integração com sistema atual
# -----------------------------

def enhance_field_validation(field_name: str, field_value: str, document_type: str, context: dict = None) -> dict:
    """
    Aplica validadores específicos baseados no campo e tipo de documento
    Integra com o sistema de alta precisão existente
    """
    context = context or {}
    
    result = {
        'field_name': field_name,
        'original_value': field_value,
        'is_valid': False,
        'confidence': 0.0,
        'normalized_value': field_value,
        'validation_method': 'basic',
        'issues': [],
        'recommendations': []
    }
    
    try:
        # Date fields
        if 'date' in field_name.lower():
            field_type = 'general'
            if 'birth' in field_name.lower():
                field_type = 'birth_date'
            elif 'expiry' in field_name.lower() or 'expire' in field_name.lower():
                field_type = 'expiry_date'
            elif 'issue' in field_name.lower():
                field_type = 'issue_date'
            
            is_valid, confidence, normalized, error = validate_date_with_context(field_value, field_type)
            result.update({
                'is_valid': is_valid,
                'confidence': confidence,
                'normalized_value': normalized,
                'validation_method': 'robust_date_parser',
                'issues': [error] if error else [],
                'recommendations': ['Check date format and readability'] if error else []
            })
        
        # Passport number
        elif field_name.lower() in ['passport_number', 'passport_no']:
            nationality = context.get('nationality', '')
            is_valid, confidence, message = validate_passport_number_with_nationality(field_value, nationality)
            result.update({
                'is_valid': is_valid,
                'confidence': confidence,
                'validation_method': 'nationality_aware_passport',
                'issues': [message] if not is_valid else [],
                'recommendations': ['Verify passport number format'] if not is_valid else []
            })
        
        # USCIS Receipt Number
        elif field_name.lower() in ['receipt_number', 'case_number'] and 'i797' in document_type.lower():
            is_valid = is_valid_uscis_receipt(field_value)
            result.update({
                'is_valid': is_valid,
                'confidence': 0.99 if is_valid else 0.1,
                'validation_method': 'uscis_receipt_validator',
                'normalized_value': field_value.upper().strip(),
                'issues': [] if is_valid else ['Invalid USCIS receipt number format or prefix'],
                'recommendations': [] if is_valid else ['Verify receipt number: 3 letters + 10 digits']
            })
        
        # SSN
        elif field_name.lower() in ['ssn', 'social_security_number']:
            is_valid = is_plausible_ssn(field_value)
            result.update({
                'is_valid': is_valid,
                'confidence': 0.95 if is_valid else 0.1,
                'validation_method': 'ssn_rules_validator',
                'issues': [] if is_valid else ['Invalid SSN format or implausible number'],
                'recommendations': [] if is_valid else ['Check SSN format: XXX-XX-XXXX with valid area/group/serial']
            })
        
        else:
            # Basic validation for other fields
            if field_value and len(field_value.strip()) > 0:
                result.update({
                    'is_valid': True,
                    'confidence': 0.8,
                    'validation_method': 'basic_presence_check'
                })
    
    except Exception as e:
        result.update({
            'is_valid': False,
            'confidence': 0.0,
            'issues': [f'Validation error: {str(e)}'],
            'validation_method': 'error'
        })
    
    return result


# Test functions for verification
def run_validation_tests():
    """Executa testes dos validadores para verificação"""
    
    # Test date normalization
    assert normalize_date("12/05/2025", prefer_day_first=True) == "2025-05-12"
    assert normalize_date("May 12, 2025") == "2025-05-12"
    assert normalize_date("D/S") == "D/S"
    
    # Test receipt validation
    assert is_valid_uscis_receipt("SRC1234567890") is True
    assert is_valid_uscis_receipt("ABC000") is False
    assert is_valid_uscis_receipt("MSC9876543210") is True
    
    # Test SSN validation
    assert is_plausible_ssn("123-45-6789") is True
    assert is_plausible_ssn("000-12-3456") is False
    assert is_plausible_ssn("666-12-3456") is False
    assert is_plausible_ssn("900-12-3456") is False
    
    print("✅ All validation tests passed!")

if __name__ == "__main__":
    run_validation_tests()

File: frontend/src/utils/validators.js
// validators.js
// Versões JS dos quatro validadores (sem libs externas).

// -----------------------------
// 1) Normalizador de datas
// -----------------------------
export function normalizeDate(s, preferDayFirst = true) {
  if (!s) return null;
  s = String(s).trim();
  const upper = s.toUpperCase();
  if (upper === "D/S" || upper === "DS") return "D/S";

  const tryParse = (str, fmts) => {
    for (const f of fmts) {
      const d = parseWithFormat(str, f);
      if (d) return toISO(d);
    }
    return null;
  };

  // Suporte simples a formatos comuns
  const df = ["DD/MM/YYYY", "DD-MM-YYYY", "DD.MM.YYYY"];
  const md = ["MM/DD/YYYY", "MM-DD-YYYY", "MM.DD.YYYY"];
  const txt = ["Mon DD, YYYY", "Month DD, YYYY"];

  const seq = preferDayFirst ? [...df, ...md, ...txt] : [...md, ...df, ...txt];

  // ISO direto?
  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return s;

  return tryParse(s, seq);
}

// Auxiliares de data (sem libs externas)
function parseWithFormat(str, fmt) {
  const months = {
    Jan:1, Feb:2, Mar:3, Apr:4, May:5, Jun:6,
    Jul:7, Aug:8, Sep:9, Oct:10, Nov:11, Dec:12
  };
  const monthNames = {
    January:1, February:2, March:3, April:4, May:5, June:6,
    July:7, August:8, September:9, October:10, November:11, December:12
  };
  let m, d, y;

  const num = (x)=>Number(x);

  switch(fmt){
    case "DD/MM/YYYY":
      m = str.match(/^(\d{2})\/(\d{2})\/(\d{4})$/); if(!m) break;
      d=num(m[1]); const mm=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, mm-1, d));
    case "DD-MM-YYYY":
      m = str.match(/^(\d{2})\-(\d{2})\-(\d{4})$/); if(!m) break;
      d=num(m[1]); const mm2=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, mm2-1, d));
    case "DD.MM.YYYY":
      m = str.match(/^(\d{2})\.(\d{2})\.(\d{4})$/); if(!m) break;
      d=num(m[1]); const mm3=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, mm3-1, d));
    case "MM/DD/YYYY":
      m = str.match(/^(\d{2})\/(\d{2})\/(\d{4})$/); if(!m) break;
      const mm4=num(m[1]); d=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, mm4-1, d));
    case "MM-DD-YYYY":
      m = str.match(/^(\d{2})\-(\d{2})\-(\d{4})$/); if(!m) break;
      const mm5=num(m[1]); d=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, mm5-1, d));
    case "MM.DD.YYYY":
      m = str.match(/^(\d{2})\.(\d{2})\.(\d{4})$/); if(!m) break;
      const mm6=num(m[1]); d=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, mm6-1, d));
    case "Mon DD, YYYY":
      m = str.match(/^([A-Z][a-z]{2})\s(\d{1,2}),\s(\d{4})$/); if(!m) break;
      d=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, (months[m[1]]||0)-1, d));
    case "Month DD, YYYY":
      m = str.match(/^([A-Z][a-z]+)\s(\d{1,2}),\s(\d{4})$/); if(!m) break;
      d=num(m[2]); y=num(m[3]); return new Date(Date.UTC(y, (monthNames[m[1]]||0)-1, d));
  }
  return null;
}

function toISO(d) {
  if (!d || Number.isNaN(d.getTime())) return null;
  return d.toISOString().slice(0,10);
}


// -----------------------------
// 2) MRZ TD3 (duas linhas, 44 chars) + checksums
// -----------------------------
const MRZ_WEIGHTS = [7,3,1];

function mrzCharValue(c){
  if (/[0-9]/.test(c)) return c.charCodeAt(0)-48;
  if (/[A-Z]/.test(c)) return c.charCodeAt(0)-55; // A=10
  if (c === "<") return 0;
  return 0;
}
function mrzChecksum(field){
  let total = 0;
  for (let i=0;i<field.length;i++){
    total += mrzCharValue(field[i]) * MRZ_WEIGHTS[i%3];
  }
  return total % 10;
}

export function parseMrzTd3(line1, line2){
  if (!line1 || !line2 || line1.length!==44 || line2.length!==44) return null;

  const number = line2.slice(0,9);
  const numberC = line2[9];
  const nationality = line2.slice(10,13);
  const dob = line2.slice(13,19);
  const dobC = line2[19];
  const sex = line2[20];
  const exp = line2.slice(21,27);
  const expC = line2[27];
  const optional = line2.slice(28,43);
  const compositeC = line2[43];

  if (mrzChecksum(number) !== Number(numberC)) return null;
  if (mrzChecksum(dob) !== Number(dobC)) return null;
  if (mrzChecksum(exp) !== Number(expC)) return null;

  const composite = number + numberC + nationality + dob + dobC + sex + exp + expC + optional;
  if (mrzChecksum(composite) !== Number(compositeC)) return null;

  const rawNames = line1.slice(5).replace(/<+$/,"");
  const parts = rawNames.split("<<").filter(Boolean);
  const surname = (parts[0]||"").replace(/</g," ").trim();
  const given = (parts[1]||"").replace(/</g," ").trim();

  const yymmddToIso = (s)=>{
    const yy = Number(s.slice(0,2));
    const mm = Number(s.slice(2,4));
    const dd = Number(s.slice(4,6));
    const year = yy >= 50 ? 1900+yy : 2000+yy;
    const d = new Date(Date.UTC(year, mm-1, dd));
    return toISO(d);
  };

  return {
    passport_number: number.replace(/</g,""),
    nationality,
    date_of_birth: yymmddToIso(dob),
    expiry_date: yymmddToIso(exp),
    sex,
    surname,
    given_names: given
  };
}


// -----------------------------
// 3) USCIS Receipt (I-797)
// -----------------------------
const VALID_PREFIXES = new Set(["EAC","WAC","LIN","SRC","MSC","IOE","NBC","NSC","TSC","VSC","YSC"]);
const receiptRx = /\b([A-Z]{3})(\d{10})\b/;

export function isValidUscisReceipt(s){
  if (!s) return false;
  const m = String(s).trim().toUpperCase().match(receiptRx);
  if (!m) return false;
  return VALID_PREFIXES.has(m[1]);
}


// -----------------------------
// 4) SSN plausível
// -----------------------------
const ssnRx = /\b(\d{3})-(\d{2})-(\d{4})\b/;

export function isPlausibleSSN(s){
  if (!s) return false;
  const m = String(s).trim().match(ssnRx);
  if (!m) return false;
  const area = Number(m[1]), group = Number(m[2]), serial = Number(m[3]);
  if (area === 0 || group === 0 || serial === 0) return false;
  if (area === 666 || area >= 900) return false;
  return true;
}

// -----------------------------
// 5) Integração com sistema frontend
// -----------------------------

export function validateFieldClient(fieldName, fieldValue, documentType = '', context = {}) {
  /**
   * Valida campos no frontend usando os mesmos validadores do backend
   */
  
  const result = {
    fieldName,
    originalValue: fieldValue,
    isValid: false,
    confidence: 0.0,
    normalizedValue: fieldValue,
    validationMethod: 'basic',
    issues: [],
    recommendations: []
  };

  try {
    // Date fields
    if (fieldName.toLowerCase().includes('date')) {
      const normalized = normalizeDate(fieldValue, true);
      if (normalized) {
        result.isValid = true;
        result.confidence = 0.95;
        result.normalizedValue = normalized;
        result.validationMethod = 'robust_date_parser';
      } else {
        result.issues.push('Invalid date format');
        result.recommendations.push('Use format: DD/MM/YYYY, MM/DD/YYYY, or Month DD, YYYY');
      }
    }
    
    // Receipt numbers
    else if (fieldName.toLowerCase().includes('receipt') && documentType.toLowerCase().includes('i797')) {
      const isValid = isValidUscisReceipt(fieldValue);
      result.isValid = isValid;
      result.confidence = isValid ? 0.99 : 0.1;
      result.validationMethod = 'uscis_receipt_validator';
      result.normalizedValue = fieldValue.trim().toUpperCase();
      if (!isValid) {
        result.issues.push('Invalid USCIS receipt format');
        result.recommendations.push('Format should be: 3 letters + 10 digits (e.g., SRC1234567890)');
      }
    }
    
    // SSN
    else if (fieldName.toLowerCase().includes('ssn') || fieldName.toLowerCase().includes('social')) {
      const isValid = isPlausibleSSN(fieldValue);
      result.isValid = isValid;
      result.confidence = isValid ? 0.95 : 0.1;
      result.validationMethod = 'ssn_rules_validator';
      if (!isValid) {
        result.issues.push('Invalid SSN format or number');
        result.recommendations.push('Format should be: XXX-XX-XXXX with valid area/group/serial');
      }
    }
    
    // Basic validation for other fields
    else if (fieldValue && fieldValue.trim().length > 0) {
      result.isValid = true;
      result.confidence = 0.8;
      result.validationMethod = 'basic_presence_check';
    }

  } catch (error) {
    result.issues.push(`Validation error: ${error.message}`);
    result.validationMethod = 'error';
  }

  return result;
}

export function validateMRZClient(line1, line2) {
  /**
   * Valida MRZ no frontend e retorna dados estruturados
   */
  const mrzData = parseMrzTd3(line1, line2);
  
  if (mrzData) {
    return {
      isValid: true,
      confidence: 0.99,
      validationMethod: 'mrz_checksum_validation',
      extractedData: mrzData,
      issues: [],
      recommendations: []
    };
  }
  
  return {
    isValid: false,
    confidence: 0.0,
    validationMethod: 'mrz_checksum_validation',
    extractedData: null,
    issues: ['MRZ checksum validation failed'],
    recommendations: ['Ensure MRZ lines are complete and clearly readable']
  };
}

// Test functions
export function runClientValidationTests() {
  // Test date normalization
  console.assert(normalizeDate("12/05/2025", true) === "2025-05-12", "Date normalization failed");
  console.assert(normalizeDate("May 12, 2025") === "2025-05-12", "Text date normalization failed");
  
  // Test receipt validation
  console.assert(isValidUscisReceipt("SRC1234567890") === true, "Valid receipt validation failed");
  console.assert(isValidUscisReceipt("ABC000") === false, "Invalid receipt validation failed");
  
  // Test SSN validation
  console.assert(isPlausibleSSN("123-45-6789") === true, "Valid SSN validation failed");
  console.assert(isPlausibleSSN("000-12-3456") === false, "Invalid SSN validation failed");
  
  console.log("✅ All client-side validation tests passed!");
}

// Export all functions for use in components
export {
  normalizeDate,
  parseMrzTd3,
  isValidUscisReceipt,
  isPlausibleSSN
};